### DISCUSSION TITLE
Update modules registration function to allow usage of "current" cradle

### DISCUSSION THREAD

**Discussion by vkislichenko** [medusajs/medusa#4546] [Feature Requests]
In a project I'm working on I'm heavily relying on awilix scopes, I manage them, add data to them, and I need "natural" scopes "cascade" and for scope I modified to be used on resolve for registrations with SCOPED Lifetime.

https://github.com/medusajs/medusa/blob/44a0d1165a28d80fa44aae45544f67d71e2b2f64/packages/modules-sdk/src/loaders/utils/load-internal.ts#L115
but here, currently all modules are restricted to use "local" scope (created from main container)

please add an option for module definition to use current cradle instead

**Comment by adrien2p** [comment]
The objective of a module is to be completely isolated from the core. It is suppose to represent a domain or feature that works autonomously. Eventually, you can inject some dependencies that would be required, but it is more dependencies such as a logger an event bus or things like that. What module are you working on? what is the purpose of the module that pushes you to want to share the entire core container? I ask to evaluate if what you are looking for wouldn't be a plugin instead for now

Just for information, the fact that the container used by the module is containing everything was a bug and not a feature in shared mode, you can see the fix [here ](https://github.com/medusajs/medusa/pull/4611), you can also think of the module as a micro service, and your micorservices don't usually share something from the outside. Let me know if that is clear.

**Reply by vkislichenko** [reply]
> The objective of a module is to be completely isolated from the core.

I get it, but this approach "breaks" scope logic built in to awilix, making module services "hold on" to the "root" instances of dependancies.
My issue specifically is that - there is no controll for the instances of the container registrations (services) after initial "boot". Awilix is a good solution, but inability to "scope"/extend stuff - nececitates really ugly overrides (like whole js module instead of specific "function" of the class inside of it)

Alternative solution would be some kind of  "scope" logic on "medusa" side (like in awilix), so that all registered containers could be reinitialized properly using provided overriden serivices instances.

**Comment by vkislichenko** [comment]
>  What module are you working on

not a specific module, it's "tenancy"

tenant containers created on boot (scoped from existing root container), on request (after tenant is resolved by a middleware) appropriate tenant container is picked up in to a request object.
each tenant container has "manager" overriden to use appropriate db connection, "event bus" is also initiated per tenant (+ few other things are added)

I had to override a lot of things, and after medusa-extender lost compatibility even more things (admin and storefront too)
